### ISSUE TITLE
Better experience when the url returns something besides what we expect

### ISSUE THREAD

**Issue by sayedihashimi** [dotnet/HttpRepl#367]
I issued a `GET` command that pointed to a `.nupkg`. The result was that a bunch of random characters were written to the console.

We should have a better experience for unknown response types.

@bradygaster @tlmii

**Comment by tlmii** [comment]
I definitely want to make this change, maybe even sneaking it into 3.1 depending on what route we go. I think the biggest question in my mind is do have an allow list (e.g. only display text/plain, application/json, application/xml, etc), a block list (would probably be too long) or some type of heuristic to determine if the contents are human-readable? I worry a little bit about the allow list being too restrictive. Perhaps the allow list is a setting the user can add to? Or maybe there's an override switch on the various HTTP commands to force display if we wouldn't normally?

**Comment by sayedihashimi** [comment]
I'm not sure what the best approach is here. Have we looked at how other similar tools handle this situation?

**Comment by bradygaster** [comment]
i'd be okay if it did one of the following:

1. Show a cue to the user, something like [binary data]
1. Download the binary stream as a file and drop it into the directory where the user is running the tool
1. Download the binary and show a file save dialog, allowing the user to choose where to drop the file

3rd option is obviously the most intense and far-reaching for a customer. It is what I would **want** to see as a customer.

**Comment by tlmii** [comment]
@sayedihashimi 
> Have we looked at how other similar tools handle this situation?

Two examples:

httpie displays a banner if you attempt to download binary data and does not display the binary data. It also has a `--download|-d` option you can specify that will download the response into a local file. They also have, effectively, an allow list (via a set of converters for specific mime types).

curl will also display a banner for binary data. It has the `--output -` option to tell it to send it to the terminal anyway and the --output <FILE>` option to save it to a file. I'm a little less clear on how they do their detection - on first glance it appears more heuristic based than mime-type based.

**Comment by bradygaster** [comment]
Curious if our friend @robpaveza (who may be a little delayed as he's oof atm) has an opinion or experience in this area in his development of the Edge Network Console tool. Seems like he would have run into a situation where he's getting binary content from an API at some point during his testing.

**Comment by robpaveza** [comment]
We don't have a way to detect binary vs. textual content yet. However, we cap displaying the raw data by default at 64kb. You should be able to detect if you're going to exceed, say, more than a couple of screens worth of text content, and then not display it by default.

**Comment by tlmii** [comment]
I think we need to make a call on whether to split this into short term and long term solutions (and either hold 3.1 for the short term, or create a 3.2 that includes it) or just move it to the 5.0 milestone. Short term could just be trimming the spew at some hard-coded length so we don't let users get into a bad state.

**Comment by bradygaster** [comment]
agreed that we table this without a clear decision on how to resolve this overall experience issue. we need to address the large-response issue separately.

**Comment by sayedihashimi** [comment]
**Thank you for your interest in HttpRepl.**

This project has been deprecated, and the repository is now archived. As part of this transition, we are closing all open issues.
Please read the https://github.com/dotnet/HttpRepl/issues/701 for alternative recommendations.
